From e7150ba2a25d973a1e4722aa2c645cdab95044de Mon Sep 17 00:00:00 2001
From: "sunny.luo" <sunny.luo@amlogic.com>
Date: Tue, 29 Jul 2014 15:44:35 +0800
Subject: [PATCH 4652/5965] PD#95060:sardadc: add saradc suspend & resume
 function for optimize consumption from Sunny.luo

Conflicts:

	drivers/amlogic/input/saradc/saradc.c
---
 drivers/amlogic/input/saradc/saradc.c | 28 ++++++++++++++++++++++++---
 1 file changed, 25 insertions(+), 3 deletions(-)

diff --git a/drivers/amlogic/input/saradc/saradc.c b/drivers/amlogic/input/saradc/saradc.c
index c045996810ca..ecd07aa7e5c9 100755
--- a/drivers/amlogic/input/saradc/saradc.c
+++ b/drivers/amlogic/input/saradc/saradc.c
@@ -306,7 +306,7 @@ static ssize_t saradc_print_flag_store(struct class *cla, struct class_attribute
 #ifndef CONFIG_MESON_CPU_TEMP_SENSOR
 static ssize_t saradc_temperature_store(struct class *cla, struct class_attribute *attr, const char *buf, size_t count)
 {
-    //int ret;
+    int ret;
 		u8 tempsen;
 		sscanf(buf, "%d", (int*)&tempsen);
 		if (tempsen) {
@@ -454,6 +454,28 @@ err_free_mem:
 	return err;
 }
 
+static int saradc_suspend(struct platform_device *pdev)
+{
+	printk("%s: disable SARADC\n", __func__);
+#if MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8
+	disable_bandgap();
+#endif
+	disable_adc();
+	disable_clock();
+	return 0;
+}
+
+static int saradc_resume(struct platform_device *pdev)
+{
+	printk("%s: enable SARADC\n", __func__);
+	enable_clock();
+	enable_adc();
+#if MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8
+	enable_bandgap();
+#endif
+	return 0;
+}
+
 static int __exit saradc_remove(struct platform_device *pdev)
 {
 	struct saradc *saradc = platform_get_drvdata(pdev);
@@ -475,8 +497,8 @@ static const struct of_device_id saradc_dt_match[]={
 static struct platform_driver saradc_driver = {
 	.probe      = saradc_probe,
 	.remove     = saradc_remove,
-	.suspend    = NULL,
-	.resume     = NULL,
+	.suspend    = saradc_suspend,
+	.resume     = saradc_resume,
 	.driver     = {
 		.name   = "saradc",
 		.of_match_table = saradc_dt_match,
-- 
2.19.0

